README.md
# PoolDilutionR

<!-- badges: start -->
[![R-CMD-check](https://github.com/COMPASS-DOE/PoolDilutionR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/COMPASS-DOE/PoolDilutionR/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

An R package for easy calculation of rates from isotope pool dilution experiments.

To load the package, type `library(PoolDilutionR)`.

## Overview

This code calculates gross production and consumption from pool dilution time series data.
Target pool size (e.g., methane, nitrate) and isotopic composition (e.g., 13C, 15N) are measured from a closed system over time, the model optimizes production rate (P) and the first order rate constant (k), by minimizing error in the model-predicted total pool size (unit volume or mass), as well as the isotopic signature.
The optimization weighs the signal:noise ratio of pool and signatures using measurement precision as well as the magnitude of change over time.
The calculations used here are based on von Fischer and Hedin 2002, 10.1029/2001GB001448, with some modifications (manuscript link pending).

Original code was written for data collected on a Picarro CRDS equipped with a SSIM2 unit.

## Inputs

Time series data of pool and isotopic composition.

![image](https://user-images.githubusercontent.com/89651444/205718596-db983482-6d38-45d7-9b8c-2ee6fd6948f5.png)

## Functions

Prediction of P, gross production rate, and k, first order rate constant for consumption.

    `pred <- pdr_predict(time = dat$time_days,
                          m0 = dat$cal12CH4ml[1] + dat$cal13CH4ml[1],
                          n0 = dat$cal13CH4ml[1],
                          P = P,
                          k = result$par["k"],
                          pool = "CH4")`
    

Optimization of predictions using data quality and deviation in pool size (Nm) and isotopic composition (Nd) over time.

    `result <- pdr_optimize(time = dat$time_days,
                          m = dat$cal12CH4ml + dat$cal13CH4ml,
                          n = dat$cal13CH4ml,
                          Nm = dat$Nm,
                          Nd = dat$Nd,
                          pool = "CH4",
                          params = params)`
                     
## Output

Optimized rates based on total pool size and atom percent composition.

![image](https://user-images.githubusercontent.com/89651444/205720528-ae4554f0-0b37-4b19-8bcc-8d74d7c956c4.png)
![image](https://user-images.githubusercontent.com/89651444/205720560-182a6316-c063-4fcb-b333-768c0f590a38.png)




NEWS.md
# PoolDilutionR 1.0.0

* Initial release of `PoolDilutionR` (2023-02-06)


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(PoolDilutionR)

test_check("PoolDilutionR")


tests/testthat/test-misc.R
test_that("Default P and k functions work", {
  expect_error(frac_P_default("bad pool"), "pool is not listed")
  expect_error(frac_k_default("bad pool"), "pool is not listed")

  p <- frac_P_default("CH4")
  expect_type(p, "double")
  expect_length(p, 1L)

  k <- frac_k_default("CH4")
  expect_type(k, "double")
  expect_length(k, 1L)
})

test_that("pdr_fractionation table is valid", {
  # There should be a maximum of one default value for each pool
  for (p in unique(pdr_fractionation$Pool)) {
    pdr_f <- subset(pdr_fractionation, pdr_fractionation$Pool == p)
    expect_lt(sum(pdr_f$Default), 2)
  }
})


tests/testthat/test-AP_pred.R
test_that("pdr_predict works", {
  # Bad inputs
  expect_error(pdr_predict(1), regexp = "must be zero")
  expect_error(pdr_predict("1"), regexp = "must be numeric")
  expect_error(pdr_predict(c(0, 2, 1)), regexp = "must increase")

  # Basic behavior, P < C
  times <- 0:5
  x <- pdr_predict(time = 0:5, m0 = 10, n0 = 1, P = 0.5, k = 0.3)
  expect_identical(nrow(x), length(times))
  expect_identical(names(x), c("mt", "nt", "AP_pred"))
  expect_true(all(diff(x$mt) < 0)) # decreasing mt
  expect_true(all(diff(x$nt) < 0)) # decreasing nt
  expect_true(all(diff(x$AP_pred) < 0)) # decreasing AP_pred

  # Basic behavior, P > C
  times <- 0:5
  x <- pdr_predict(time = 0:5, m0 = 10, n0 = 1, P = 5, k = 0.3)
  expect_identical(nrow(x), length(times))
  expect_identical(names(x), c("mt", "nt", "AP_pred"))
  expect_true(all(diff(x$mt) > 0)) # increasing mt
  expect_true(all(diff(x$nt) < 0)) # decreasing nt
  expect_true(all(diff(x$AP_pred) < 0)) # decreasing AP_pred

  # Behavior: No production or consumption
  # If inputs P = k = 0.0 , then outputs mt and nt should be constant
  # and equal to m0 and n0 respectively.


  # Behavior: No methane and no production
  # A variant of the previous one. If inputs P = m0 = n0 = 0.0 ,
  # then outputs mt and nt should be constant zero
  x <- pdr_predict(time = times, m0 = 0, n0 = 0, P = 0.0, k = 0.1)
  expect_true(all(x$mt == 0))
  expect_true(all(x$nt == 0))

  # Behavior: No discrimination
  # If inputs frac_P = frac_k = 1.0 , then output AP_pred should be constant.
})


R/data.R
#' P and k fractionation values
#'
#' A compendium of possible production (P) and consumption (k) fractionation
#' values, by pool.
#'
#' @format
#' \describe{
#'   \item{Pool}{Name of pool (gas or solid)}
#'   \item{frac_P}{Fractionation value of production (P)}
#'   \item{frac_k}{Fractionation value of consumption (k)}
#'   \item{Default}{Default for this pool? Logical}
#'   \item{Source}{Source paper or URL}
#' }
#' @note Currently there is only one set of fractionation values available,
#' from von Fischer and Hedin (2002, 10.1029/2001GB001448).
"pdr_fractionation"

#' Example time series data from a methane dilution pool experiment.
#'
#' Sequential measurements of methane concentration and isotopic signature were taken using a Picarro G2920 with a Small Sample Introduction module.
#' This instrument provides gas concentrations in ppm and signatures in delta-13C, here we provide those data converted into volume of methane and atom percent.
#'
#' @format
#' \describe{
#'   \item{id}{Sample ID, a factor}
#'   \item{time_days}{time in days between measurements, starting at 0}
#'   \item{cal12CH4ml}{ml of 12C-CH4 at each timestep}
#'   \item{cal13CH4ml}{ml of 13C-CH4 at each timestep}
#'   \item{AP_obs}{atom percent 13C-CH4 at each timestep}
#' }
"Morris2023"


R/misc.R
utils::globalVariables("pdr_fractionation")

#' Retrieve default P fractionation value for a pool
#'
#' @param pool Name of pool, character
#'
#' @return The default entry for \code{pool} listed
#' in \code{\link{pdr_fractionation}}.
#' @export
#'
#' @examples
#' frac_P_default("CH4")
frac_P_default <- function(pool) {
  if (!pool %in% pdr_fractionation$Pool) {
    stop("This pool is not listed in pdr_fractionation")
  }
  pdr_fractionation[
    pdr_fractionation$Pool == pool &
      pdr_fractionation$Default,
    "frac_P"
  ]
}


#' Retrieve default k fractionation value for a pool
#'
#' @param pool Name of pool, character
#'
#' @return The default entry for \code{pool} listed
#' in \code{\link{pdr_fractionation}}.
#' @export
#'
#' @examples
#' frac_k_default("CH4")
frac_k_default <- function(pool) {
  if (!pool %in% pdr_fractionation$Pool) {
    stop("This pool is not listed in pdr_fractionation")
  }
  pdr_fractionation[
    pdr_fractionation$Pool == pool &
      pdr_fractionation$Default,
    "frac_k"
  ]
}


R/AP_pred.R
#' Predict total pool, heavy isotope pool, and atom percent
#'
#' @param time Vector of numeric time values (e.g. days); first should be zero
#' @param m0 total pool size at time zero, as a volume
#' @param n0 pool size of heavy isotope at time zero, as a volume
#' @param P production rate, unit gas/unit time
#' @param k first-order rate constant for consumption, 1/unit time
#' @param pool Name of pool; see \code{\link{pdr_fractionation}}
#' @param frac_P Fractionation value for production; see \code{\link{pdr_fractionation}}
#' @param frac_k Fractionation value for consumption; see \code{\link{pdr_fractionation}}
#'
#' @return Returns a data frame with \code{mt}, \code{nt}, and \code{AP_pred} (atom percent) for each time step
#' @export
#'
#' @note This is Eq. 11 from von Fischer and Hedin 2002 with a few modifications.
#' @author K.A. Morris & B. Bond-Lamberty
#' @examples
#' pdr_predict(time = 0:5, m0 = 10, n0 = 1, P = 0.5, k = 0.3)
pdr_predict <- function(time, m0, n0, P, k,
                        pool = "CH4",
                        frac_P = frac_P_default(pool),
                        frac_k = frac_k_default(pool)) {
  if (!is.numeric(time)) stop("Time must be numeric.")
  if (!all(diff(time) > 0)) stop("Time values must increase.")
  if (time[1] != 0.0) stop("First time value must be zero.")

  kfrac <- k * frac_k
  pfrac <- P * frac_P
  nt <- pfrac / kfrac - (pfrac / kfrac - n0) * exp(-kfrac * time)

  # Equation 5 (and denominator in Eq. 11 of von Fischer and Hedin 2002):
  mt <- P / k - (P / k - m0) * exp(-k * time)
  ap <- (nt / mt) * 100
  # If any nt values are zero, we want the corresponding atom percent values
  # also to be zero, not Inf or Nan
  ap[nt == 0] <- 0.0

  data.frame(mt = mt, nt = nt, AP_pred = ap)
}


R/cost_function.R
#' Cost function between observed and predicted pools
#'
#' @param params Named list holding optimizer-assigned values for parameters
#' @param time Vector of numeric time values; first should be zero
#' @param m Observed total pool size, same length as time
#' @param n Observed pool size of heavy isotope, same length as time
#' @param m_prec Instrument precision for pool size, expressed as a standard deviation
#' @param ap_prec Instrument precision for atom percent, expressed as a standard deviation
#' @param P production rate, unit pool size/unit time
#' @param k first-order rate constant for consumption, 1/unit time
#' @param pool Name of pool; see \code{\link{pdr_fractionation}}
#' @param frac_P Fractionation value for production; see \code{\link{pdr_fractionation}}
#' @param frac_k Fractionation value for consumption; see \code{\link{pdr_fractionation}}
#' @param log_progress An optional logging function
#'
#' @importFrom stats sd
#' @return Returns a cost metric summarizing the difference between the
#' predicted and observed \code{m} (total pool size) and \code{AP} (atom percent).
#' @export
#'
#' @note This implements Equations 12-14 from von Fischer and Hedin (2002).
#' @author K.A. Morris & B. Bond-Lamberty
#' @examples
#' m <- c(10, 8, 6, 5, 4, 3)
#' n <- c(1, 0.7, 0.6, 0.4, 0.3, 0.2)
#' pdr_cost(params = list(P = 0.5, k = 0.3), time = 0:5, m, n, m_prec = 0.001, ap_prec = 0.01)
pdr_cost <- function(params, # values are set by optim()
                     time, m, n, m_prec, ap_prec,
                     P,
                     k,
                     pool = "CH4",
                     frac_P = frac_P_default(pool),
                     frac_k = frac_k_default(pool),
                     log_progress = NULL) {
  # If the optimizer passes in values for P/k/frac_P/frac_k, they
  # override the defaults (named parameters)
  if ("P" %in% names(params)) P <- params[["P"]]
  if ("k" %in% names(params)) k <- params[["k"]]
  if ("frac_P" %in% names(params)) frac_P <- params[["frac_P"]]
  if ("frac_k" %in% names(params)) frac_k <- params[["frac_k"]]

  pred <- pdr_predict(
    time = time,
    m0 = m[1],
    n0 = n[1],
    P = P,
    k = k,
    frac_P = frac_P,
    frac_k = frac_k
  )

  ap_obs <- (n / m) * 100
  pool_weight <- sd(m) / m_prec # Normalization factor for pool size, see Eq. 12
  ap_weight <- sd(ap_obs) / ap_prec # Normalization factor for isotopic signature, see Eq. 13

  # von Fischer and Hedin (2002) equation 14
  cost <- sum(abs(ap_obs - pred$AP_pred) / sd(ap_obs)) * ap_weight +
    sum(abs(m - pred$mt) / sd(m)) * pool_weight

  # Log progress and return to optimizer
  if (!is.null(log_progress)) {
    log_progress(data.frame(P = P, k = k, frac_P = frac_P, frac_k = frac_k, cost = cost))
  }

  cost
}


R/optimize.R
#' Optimize production and consumption parameters for pool dilution data
#'
#' @param time Vector of numeric time values (e.g. days); first should be zero
#' @param m Observed total pool size (as a volume), same length as time
#' @param n Observed heavy isotope (as a volume), same length as time
#' @param m_prec Instrument precision for pool size, expressed as a standard deviation
#' @param ap_prec Instrument precision for atom percent, expressed as a standard deviation
#' @param P production rate, unit gas/unit time
#' @param k first-order rate constant for consumption, 1/unit time
#' @param params_to_optimize Named vector of parameters ("P", "k", "frac_P",
#' and/or "frac_k") to optimize against observations
#' @param pool Name of pool to use when looking up fractionation values if they
#' are not supplied; see \code{\link{pdr_fractionation}}
#' @param frac_P Fractionation value for production; see \code{\link{pdr_fractionation}}
#' @param frac_k Fractionation value for consumption; see \code{\link{pdr_fractionation}}
#' @param other_params Other parameters pass on to \code{\link{optim}}
#' @param cost_fn Cost function to use; the default is \code{\link{pdr_cost}}
#' @param prediction_fn Prediction function that the cost function will use;
#' the default is \code{\link{pdr_predict}}
#' @param include_progress Include detailed optimizer progress data in output?
#' @param quiet Suppress output messages, logical
#' @note Currently there is only one set of fractionation values available in
#' \code{\link{pdr_fractionation}}, from von Fischer and Hedin
#' (2002, 10.1029/2001GB001448).
#' @importFrom stats optim
#' @return The output of \code{\link{optim}}.
#' @seealso \code{\link{pdr_optimize_df}}
#' @export
#'
#' @examples
#' tm <- 0:5
#' m <- c(10, 8, 6, 5, 4, 3)
#' n <- c(1, 0.7, 0.6, 0.4, 0.3, 0.2)
#' m_prec <- 0.001
#' ap_prec <- 0.01
#'
#' # Optimize values for P (production) and k (consumption), provide starting values for P and k
#' pdr_optimize(time = tm, m, n, m_prec, ap_prec, P = 0.5, k = 0.3)
#' # If we don't provide a value for k, it can be estimated from the data
#' pdr_optimize(tm, m, n, m_prec, ap_prec, P = 0.5)
#' # Hold k and frac_k constant (ie., k = estimated k0, frac_k = default value), optimize P and frac_P
#' pdr_optimize(tm, m, n, m_prec, ap_prec, P = 0.5, params_to_optimize = c("P", "frac_P"))
#' # Optimize only k (provide P and exclude from params_to_optimize)
#' pdr_optimize(tm, m, n, m_prec, ap_prec, P = 0.5, params_to_optimize = "k")
#' # Optimize only k, bounding its possible values
#' op <- list(lower = c("k" = 0.2), upper = c("k" = 0.3))
#' pdr_optimize(tm, m, n, m_prec, ap_prec, 0.5, 0.27, params_to_optimize = "k", other_params = op)
pdr_optimize <- function(time, m, n, m_prec, ap_prec,
                         P,
                         k,
                         params_to_optimize = c("P", "k"),
                         pool = "CH4",
                         frac_P = NULL,
                         frac_k = NULL,
                         other_params = list(),
                         cost_fn = pdr_cost,
                         prediction_fn = pdr_predict,
                         include_progress = FALSE,
                         quiet = FALSE) {
  if (is.null(frac_P)) {
    if (!quiet) message("No frac_P provided; looking up from pdr_fractionation table")
    frac_P <- frac_P_default(pool)
  }
  if (is.null(frac_k)) {
    if (!quiet) message("No frac_k provided; looking up from pdr_fractionation table")
    frac_k <- frac_k_default(pool)
  }

  # Set defaults if not given by user
  other_params <- set_default_params(other_params)

  # Create a closure for logging progress
  log_msgs <- list()
  step <- 1
  plog <- function(x) {
    log_msgs[[step]] <<- as.data.frame(x)
    step <<- step + 1
  }

  # Estimate k starting value if not given
  if (missing(k)) {
    k <- pdr_estimate_k0(time, n, frac_k, quiet = quiet)
  }

  # Create the optim's 'par' vector that controls which parameters to optimize
  all_params <- c("P" = P, "k" = k, "frac_P" = frac_P, "frac_k" = frac_k)
  if (any(!params_to_optimize %in% names(all_params))) {
    stop("params_to_optimize must be P, k, frac_P, and/or frac_k")
  }
  params <- all_params[params_to_optimize]

  # Make sure 'lower' and 'upper' only have the variables being optimized;
  # otherwise bad things happen inside of optim()
  other_params$lower <- other_params$lower[intersect(
    names(params),
    names(other_params$lower)
  )]
  other_params$upper <- other_params$upper[intersect(
    names(params),
    names(other_params$upper)
  )]

  # Call optim()
  out <- optim(
    par = params,
    fn = cost_fn,
    method = other_params[["method"]],
    lower = other_params[["lower"]],
    upper = other_params[["upper"]],
    control = other_params[["control"]],

    # "..." that the optimizer will pass to cost_fn:
    time = time,
    m = m,
    n = n,
    m_prec = m_prec,
    ap_prec = ap_prec,
    P = P,
    k = k,
    frac_P = frac_P,
    frac_k = frac_k,
    log_progress = plog
  )

  out$initial_par <- all_params
  out$initial_other <- other_params
  if (include_progress) {
    out$progress <- do.call(rbind, c(log_msgs, make.row.names = FALSE))
  }
  out
}


# Set default parameters for optimization
#
# @param other_params A named list of parameters
#
# @return The \code{params} list with entries filled in as needed.
set_default_params <- function(other_params) {
  # If method not given, set default
  if (is.null(other_params[["method"]])) {
    other_params[["method"]] <- "L-BFGS-B"
  }
  # If bounds not given, provide some so that the optimizer
  # isn't allowed to produce <0 values for P, nor <=0 for k
  if (is.null(other_params[["lower"]])) {
    other_params[["lower"]] <- c("P" = 0.0, "k" = 0.0001, "frac_P" = 0, "frac_k" = 0.0001)
  }
  if (is.null(other_params[["upper"]])) {
    other_params[["upper"]] <- c("P" = Inf, "k" = Inf, "frac_P" = 1, "frac_k" = 1)
  }
  # If control not given, use an empty list
  if (is.null(other_params[["control"]])) {
    other_params[["control"]] <- list()
  }
  other_params
}


#' Estimate initial k from heavy isotope concentration data
#'
#' @param time Vector of numeric time values (e.g. days); first should be zero
#' @param n Observed heavy isotope (as a volume), same length as time
#' @param frac_k Fractionation: 13C consumption as a fraction of 12C consumption
#' @param quiet Suppress output message, logical
#'
#' @importFrom stats lm
#' @return Initial estimate of k0 (consumption rate constant)
#' @export
#'
#' @examples
#' pdr_estimate_k0(1:5, c(1, 0.9, 0.7, 0.65, 0.4), frac_k = 0.98)
pdr_estimate_k0 <- function(time, n, frac_k, quiet = FALSE) {
  if (!all(diff(time) > 0)) stop("Time values must increase.")

  # Estimate starting k by slope of 13C following para. 21 in VfH2002:
  mod <- lm(log(n) ~ time)
  n_slope <- unname(mod$coefficients["time"])
  # Generally, this slope is negative (net 13CH4 consumption)
  # If not, our k0 estimate below won't work
  # For now, ensure this is true; there's probably a more sophisticated
  # way to estimate k0 in this case but save that for the future
  n_slope <- -abs(n_slope)

  # "...multiplied by 1/a to correct for fractionation against the
  # labeled methane." (BBL: this should be "1/-a"; see equation 8)
  k0 <- n_slope / -frac_k
  if (!quiet) message("Estimated k0 = ", round(k0, 3), " from n_slope = ", round(n_slope, 3))

  k0
}


#' Optimize production and consumption parameters for pool dilution data
#'
#' @param ... Parameters to be passed on to \code{\link{pdr_optimize}}
#'
#' @return The output of \code{\link{pdr_optimize}} summarized in a data frame,
#' with one line per parameter estimates (\code{P}, \code{k},
#' \code{frac_P}, and/or \code{frac_k}).
#' @export
#' @seealso \code{\link{pdr_optimize}}
#'
#' @examples
#' tm <- 0:5
#' m <- c(10, 8, 6, 5, 4, 3)
#' n <- c(1, 0.7, 0.6, 0.4, 0.3, 0.2)
#' m_prec <- 0.001
#' ap_prec <- 0.01
#' # Optimize values for P (production) and k (consumption)
#' pdr_optimize_df(time = tm, m, n, m_prec, ap_prec, P = 0.5, k = 0.3)
pdr_optimize_df <- function(...) {
  x <- pdr_optimize(...)

  # Return results as a 'tidy' data.frame
  initials <- data.frame(as.list(x$initial_par))
  row.names(initials) <- NULL
  initials <- initials[rep(1, length(x$par)), ]
  colnames(initials) <- paste0("initial.", colnames(initials))

  out1 <- data.frame(
    par = names(x$par),
    value = unname(x$par)
  )
  out2 <- data.frame(convergence = rep(x$convergence, nrow(out1)))

  cbind(out1, initials, out2)
}
